### PR TITLE
build package artifacts for Linux with CPack

### PR DESCRIPTION
**Pull request by qrkourier**
This PR includes a CPack implementation for CMake and an Actions workflow `cpack.yml` that builds a package artifact for a handful of mainstream Linux distributions on `x86_64`. The pre-existing CI continues to build `ziti-edge-tunnel` for the same target OSs and ARCHs.

The goals of this PR include:
* Deliver a CMake-based build system for Linux distribution packages
* Deliver on-demand CI that composes the package artifact, and will be integrated with future distribution channels (e.g `yum` and `apt` repos in Artifactory)

The changes to the top-level CMakeLists.txt files follow a rationale: If building packages then require a semver from Git (not the dummy version like 0.0.0). You may pass an explicit version to use or override the requirement for a semver with command-line build options. The semver is important for package managers like `yum` and `apt` to compute upgrade and downgrade actions based upon the semver that's embedded in the package.

CPack builds a `deb` for the Debian family and `rpm` for RedHat based upon the OS detected by CMake. These packages install systemd specific units, and are intended to be run on mainline distributions which use systemd as their init system.

The Actions workflow uses the matrix strategy to loop over a list of Linuxes for which parallel jobs are spawned to build the Linux package in the target OS by way of Actions `job.container` spec.

1. A runner VM Ubuntu 20.04 (latest) runs a runner container for each matrix element.

2. The runner container has the same base image as the target OS for the package (e.g. redhat-7, ubuntu-18.04) to ensure a build that is compatible for the target OS. The runner container spawns and waits for the builder container of the same OS to produce the artifact, then runs the binary artifact and installs the package artifact as a simple test.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,22 +4,69 @@ if(NOT ZITI_SDK_C_BRANCH)
...
 message("tunnel only = ${TUNNEL_SDK_ONLY}")
 
-execute_process(
-        COMMAND git describe --tags --dirty=-local
+find_package(Git)
+if(NOT GIT_VERSION AND GIT_FOUND)
+    message("Found Git executable \"${GIT_EXECUTABLE}\".")
+    # Generate a git-describe version string from Git repository tags
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} describe --tags --dirty=-local --match "v*"
```
Comment (qrkourier) [inline CMakeLists.txt:16]: only match semver tags `v*`

File: CMakeLists.txt [hunk 2]
```diff
@@ -35,10 +82,9 @@ file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/bundle)
 
 macro(BUNDLE_COMP comp)
     add_custom_target(${comp}-bundle
-            BYPRODUCTS ${CMAKE_BINARY_DIR}/bundle/${comp}-${CPACK_SYSTEM_NAME}_${CMAKE_SYSTEM_PROCESSOR}.zip
-            DEPENDS ${comp}
-            COMMAND 7z a ${CMAKE_BINARY_DIR}/bundle/${comp}-${CPACK_SYSTEM_NAME}_${CMAKE_SYSTEM_PROCESSOR}.zip ${comp}${CMAKE_EXECUTABLE_SUFFIX} ${${comp}_BUNDLE_COMPS}
-            )
+        BYPRODUCTS ${CMAKE_BINARY_DIR}/bundle/${comp}-${CMAKE_SYSTEM_NAME}_${CMAKE_SYSTEM_PROCESSOR}.zip
```
Comment (qrkourier) [inline CMakeLists.txt:85]: eliminate the need for 7zip in all build environments

**Comment by qrkourier** [comment]
To demonstrate the package build, I've triggered the on-demand (on-dispatch) CI job that builds packages for the current HEAD in [this workflow run](https://github.com/qrkourier/ziti-tunnel-sdk-c/actions/runs/2034926393). 

On that page you can see the list of downloadable package artifacts that were produced. When we decide the release procedure and deliver distribution repos for `rpm` and `deb` then those same artifacts will be uploaded to Artifactory as part of a normal release.

**Review by scareything — COMMENTED** [review]
in general this seems like a lot of repetition for each linux distro. I was thinking that we would build once on the oldest distro (kernel,glibc,toolchain) that we can tolerate and package that binary into the different packages.

File: .github/workflows/cpack.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          cat /etc/*-release
+          ls -horAS ./build/
+
+      - name: list program artifacts
+        run: |
+          cat /etc/*-release
+          ls -horAS ./build/programs/ziti-edge-tunnel/
+
+      - name: run binary artifact
+        if: ${{ matrix.distro.name != 'redhat' }} # TODO: error while loading shared libraries: libatomic.so.1
```
[the remark this answers is not in the document]
Comment (sabedevops) [reply .github/workflows/cpack.yml:100]: @qrkourier I believe this one was an oversight in the development process?

File: .github/actions/openziti-tunnel-build-action/Dockerfile.redhat-7 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+RUN yum -y install \
+        "@Development Tools" \
+        centos-release-scl \
+        doxygen \
+        graphviz \
+        python3 \
+        zlib-devel \
+    &&  yum -y install \
+        devtoolset-10 \
```
[the remark this answers is not in the document]
Comment (sabedevops) [reply .github/actions/openziti-tunnel-build-action/Dockerfile.redhat-7:24]: I was not aware we could now compile with GCC 11. Will update.

File: CMakeLists.txt [hunk 5]
```diff
@@ -4,22 +4,69 @@ if(NOT ZITI_SDK_C_BRANCH)
...
+        message(WARNING "SEMVER could not be parsed from GIT_VERSION: ${GIT_VERSION}. Setting to PROJECT_SEMVER: ${PROJECT_SEMVER}")
+    else()
+        message("Parsed GIT_VERSION: \"${GIT_VERSION}\", PROJECT_SEMVER: \"${PROJECT_SEMVER}\".")
+    endif()
+endif()
+
+option(DISABLE_SEMVER_VERIFICATION "Toggle SEMVER verification for BUILD_DIST_PACKAGES" OFF)
+option(BUILD_DIST_PACKAGES "Build packages for distribution package managers" OFF)
+
+# Require explicitely disabling SEMVER verification for bulding DIST packages.
```
[the remark this answers is not in the document]
Comment (sabedevops) [reply CMakeLists.txt:51]: Thank you for the catch. Will fix.

File: .github/actions/openziti-tunnel-build-action/Dockerfile.redhat-7 (hunk 4 above)
[the remark this answers is not in the document]
Comment (qrkourier) [reply .github/actions/openziti-tunnel-build-action/Dockerfile.redhat-7]: bumped devtoolset to 11 across the repo

File: .github/workflows/cpack.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (qrkourier) [reply .github/workflows/cpack.yml]: Good catch on the TODO. I'm working on this one now in my branch.

File: .github/workflows/cmake.yml [hunk 6]
```diff
@@ -39,25 +39,38 @@ jobs:
...
+        uses: actions/checkout@v3
         with:
           fetch-depth: 0
+
       - name: make build dir
         run: cmake -E make_directory ${{runner.workspace}}/build
 
-      - name: install toolchain
-        if: matrix.install != ''
+      - name: install Debian crossbuild toolchain
```
[the remark this answers is not in the document]
Comment (qrkourier) [reply .github/workflows/cmake.yml]: Fixed in source branch.

File: CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (qrkourier) [reply CMakeLists.txt]: Fixed in source branch.

**Comment by qrkourier** [comment]
> in general this seems like a lot of repetition for each linux distro. I was thinking that we would build once on the oldest distro (kernel,glibc,toolchain) that we can tolerate and package that binary into the different packages.

Definitely considered that minimal approach, and at first it seemed best to me too. We'll definitely need at least four permutations:
1. old redhat
2. new redhat
3. old ubuntu
4. new ubuntu

In total we have six, and certainly the vast majority of the differences are expressed in the first four. At that point it felt inexpensive to add the last two, the Dockerfiles were already written and tested, and the fact of the ability to build the artifact by no means implies that we _shall_ advertise and distribute every artifact. In fact, now that the framework (loosely termed) is developed, we can easily add and remove any OS we wish from the set of package artifacts we _can_ build. This is intentionally distinct from the question of distribution (i.e. package repos) that we will later choose to advertise. That's why I'm now less concerned about the repetition than I was at first.

File: programs/ziti-edge-tunnel/package/CPackGenConfig.cmake [hunk 7]
```diff
@@ -0,0 +1,34 @@
+cmake_policy(SET CMP0057 NEW)
+
+if(CPACK_GENERATOR MATCHES "RPM")
+    set(CPACK_RPM_BUILDREQUIRES "cmake >= ${CMAKE_MINIMUM_REQUIRED_VERSION}, systemd, gawk, gcc-c++ >= 4.9, python3, zlib-devel")
+    if(CPACK_OS_RELEASE_NAME IN_LIST CPACK_RPM_DISTRIBUTIONS AND CPACK_OS_RELEASE_VERSION VERSION_GREATER "7")
+        list(APPEND CPACK_RPM_BUILDREQUIRES "systemd-rpm-macros")
+    endif()
+    set(CPACK_RPM_PACKAGE_SOURCES OFF)
+    set(CPACK_RPM_PACKAGE_REQUIRES "iproute, sed, systemd")
```
Comment (sabedevops) [inline programs/ziti-edge-tunnel/package/CPackGenConfig.cmake:9]: @qrkourier Add `libatomic`  to list here

File: programs/ziti-edge-tunnel/package/CPackGenConfig.cmake [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    set(CPACK_RPM_PACKAGE_REQUIRES_PREUN "systemd")
+    set(CPACK_RPM_PACKAGE_REQUIRES_POSTUN "systemd")
+    #[[
+    set(CPACK_RPM_ziti-edge-tunnel_PACKAGE_NAME "${CPACK_RPM_PACKAGE_NAME}")
+    set(CPACK_RPM_ziti-edge-tunnel_FILE_NAME "RPM-DEFAULT")
+    ]]
+    endif(CPACK_GENERATOR MATCHES "RPM")
+
+if(CPACK_GENERATOR MATCHES "DEB")
+    set(CPACK_DEBIAN_PACKAGE_DEPENDS "iproute2, sed, systemd")
```
Comment (sabedevops) [inline programs/ziti-edge-tunnel/package/CPackGenConfig.cmake:32]: @qrkourier Add `libatomic1` here. Notice the `1` for this one.

**Comment by qrkourier** [comment]
Package job run for current rev with all the fixes discussed above is here: https://github.com/qrkourier/ziti-tunnel-sdk-c/actions/runs/2041485430

Ready for review!

**Comment by qrkourier** [comment]
The on-demand CI job for the current HEAD is here: https://github.com/qrkourier/ziti-tunnel-sdk-c/actions/runs/2097935179

**Review by scareything — APPROVED** [review]

**Pull request merged**